opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java

```diff
@@ -338,39 +353,97 @@ private ExprValue createOpenSearchDateType(Content value, ExprType type) {
     return new ExprTimestampValue((Instant) value.objectValue());
   }
 
-  private ExprValue parseStruct(Content content, String prefix) {
+  /**
+   * Parse struct content.
+   * @param content Content to parse.
+   * @param prefix Prefix for Level of object depth to parse.
+   * @param supportArrays Are non-object arrays supported.
+   * @return Value parsed from content.
+   */
+  private ExprValue parseStruct(Content content, String prefix, boolean supportArrays) {
     LinkedHashMap<String, ExprValue> result = new LinkedHashMap<>();
     content.map().forEachRemaining(entry -> result.put(entry.getKey(),
         parse(entry.getValue(),
             makeField(prefix, entry.getKey()),
-            type(makeField(prefix, entry.getKey())))));
+            type(makeField(prefix, entry.getKey())), supportArrays)));
     return new ExprTupleValue(result);
   }
 
   /**
-   * Todo. ARRAY is not completely supported now. In OpenSearch, there is no dedicated array type.
-   * <a href="https://opensearch.org/docs/latest/opensearch/supported-field-types/nested/">docs</a>
-   * The similar data type is nested, but it can only allow a list of objects.
+   * Parse array content. Can also parse nested which isn't necessarily an array.
+   * @param content Content to parse.
+   * @param prefix Prefix for Level of object depth to parse.
+   * @param supportArrays Are non-object arrays supported.
+   * @param type Type of content parsing.
+   * @return Value parsed from content.
    */
-  private ExprValue parseArray(Content content, String prefix) {
+  private ExprValue parseArray(
+      Content content,
+      String prefix,
+      ExprType type,
+      boolean supportArrays
+  ) {
     List<ExprValue> result = new ArrayList<>();
-    // ExprCoreType.ARRAY does not indicate inner elements type.
-    if (Iterators.size(content.array()) == 1 && content.objectValue() instanceof JsonNode) {
-      result.add(parse(content, prefix, Optional.of(STRUCT)));
+
+    // ARRAY is mapped to nested but can take the json structure of an Object.
+    if (content.objectValue() instanceof ObjectNode) {
+      result.add(parseStruct(content, prefix, supportArrays));
+      // non-object type arrays are only supported when parsing inner_hits of OS response.
+    } else if (
+        !(type instanceof OpenSearchDataType
+            && ((OpenSearchDataType) type).getExprType().equals(ARRAY))
+        && !supportArrays) {
+      return parseInnerArrayValue(content.array().next(), prefix, type, supportArrays);
```

Comment:
should this be `result.add(parseInnerArrayValue(...))`?

Reply:
Object arrays aren't necessarily supported and only the first index is parsed if the array is not of type nested.

[ObjectFieldSelectIT.java#L69-L93](https://github.com/opensearch-project/sql/blob/main/integ-test/src/test/java/org/opensearch/sql/legacy/ObjectFieldSelectIT.java#L69-L93)